spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/EntityPropertyValueProvider.java

```diff
@@ -46,27 +50,71 @@ public EntityPropertyValueProvider(BaseEntity entity, ReadWriteConversions readW
 	}
 
 	@Override
-	@SuppressWarnings("unchecked")
 	public <T> T getPropertyValue(DatastorePersistentProperty persistentProperty) {
-		String fieldName = persistentProperty.getFieldName();
+		boolean isMap = Map.class.isAssignableFrom(persistentProperty.getType());
+		TypeInformation singularType;
+		if (isMap) {
+			singularType = persistentProperty.getEmbeddedMapValueType();
+		}
+		else {
+			singularType = persistentProperty.getComponentType() == null ? null
+					: ClassTypeInformation.from(persistentProperty.getComponentType());
+		}
+		Class collectionType = persistentProperty.getType();
+		if (singularType == null) {
+			singularType = ClassTypeInformation.from(collectionType);
+			collectionType = null;
+		}
+		return getPropertyValue(persistentProperty.getFieldName(),
+				persistentProperty.isEmbedded(), isMap,
+				persistentProperty.isEmbeddedComponents(), collectionType,
+				singularType);
+	}
 
+	/**
+	 * Get a property value from the entity.
+	 * @param fieldName the name of the field to get.
+	 * @param isEmbedded if the property is an embedded entity.
+	 * @param isEmbeddedComponents if the property is collection-like and contains
+	 * embedded items.
+	 * @param collectionType the collection type if the property is not singular. null if
+	 * the property is singular.
+	 * @param componentType the singular item type.
+	 * @return the property converted from the entity.
+	 */
+	public <T> T getPropertyValue(String fieldName, boolean isEmbedded,
+			boolean isEmbeddedComponents,
+			Class collectionType, Class componentType) {
+		return (T) getPropertyValue(fieldName, isEmbedded, false, isEmbeddedComponents,
+				collectionType,
+				ClassTypeInformation.from(componentType));
+	}
+
+	@SuppressWarnings("unchecked")
+	private <T> T getPropertyValue(String fieldName, boolean isEmbedded, boolean isMap,
+			boolean isEmbeddedComponents,
+			Class collectionType, TypeInformation componentType) {
```

Comment:
Combine two collection and component type into a single type information for the entire thing. removes one arg.